### ISSUE TITLE
--fix is dangerous with quoted variables

### ISSUE THREAD

**Issue by kylec1** [rodjek/puppet-lint#437]
The "Variables standing by themselves should not be quoted" rule is reasonable, but in some cases, removing quotes alters the type and breaks the Puppet code. `--fix` should probably not perform semantics-changing fixes by default.

**Comment by rnelson0** [comment]
@kylec1 There are [control comments](http://puppet-lint.com/controlcomments/) to assist with this. Would those suffice? If you have some specific concerns, can you update the ticket with examples? Thanks, and sorry for the delay in response.

**Comment by kylec1** [comment]
Sadly, I don't think I have the example anymore; it was internal code that
has since been removed.
But I reported the issue after some puppet variables that were integers
being quoted into strings were unquoted by `--fix`, causing other puppet
code that treated data differently depending on type to fail in subtle ways
once deployed.
This caught me by surprise, as I wouldn't have expected a lint tool to
change behavior with an autofix.
Control comments would make it possible to avoid having subsequent `--fix`
runs cause the same problem once I identified it, which is nice, but I
wouldn't have known that I should've looked through our thousands of lines
of puppet code where this issue might need a control comment before running
`--fix`, so it doesn't resolve the specific issue.

I don't know if there's a practical fix aside from deciding you are or
aren't okay with the risk, with the latter probably involving a flag for
people to opt in. Choosing to only fix provably unnecessary quotes likely
requires more analysis than it might be reasonable for a linter to perform.

**Comment by rnelson0** [comment]
> Choosing to only fix provably unnecessary quotes likely requires more analysis than it might be reasonable for a linter to perform.

This is my concern, too. Having some samples would help us identify suspect code and determine the cost/benefit ratio. We haven't seen other reports of similar problems, so I am hopeful that some of the fixes over the past year may have already accounted for the worst of the problems you encountered. If it's okay with you, I'd like to close this issue out. We can re-open it or start a new issue if we find enough samples to analyze.

**Comment by kylec1** [comment]
Feel free to close at will; this isn't an issue for me anymore, so I'm
neutral.
As for examples, I recall the original being convoluted, but the pattern
was essentially:

```
$value = 1;

thing {
   config_value => "${value}",
}
```

where `thing` behaves differently for a numeric type (as defined by stdlib
`is_numeric`) than for string types.
By removing the quotes, the numeric value becomes a string value, and
behavior changes. This seems safe, and likely
is in most cases, and in my case, may have only been problematic due to an
`is_empty` bug down the call stack or something, but
that semantic change changed behavior to where a rollback was needed. If
that isn't common or worrisome, that's reasonable.
I'm happy to defer to your judgment.

**Comment by rnelson0** [comment]
I cannot speak to how common it is, but in this example, lint won't know about the type definition. I think it would be extremely unlikely that it could intuit that properly. Puppet 4 data types, or assignments like `$value = "1"` would address it just as well and gives more control to the user. Thanks for the rapid feedback after such a long wait!
